fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -458,8 +475,417 @@ Value makePadding(int size) {
+		}
+		return Void();
+	}
+
+	// Ends the current block if necessary based on bytesNeeded.
+	ACTOR static Future<Void> newBlockIfNeeded(EncryptedRangeFileWriter* self, int bytesNeeded) {
+		if (expectedFileSize(self) + bytesNeeded > self->blockEnd) {
+			wait(newBlock(self, bytesNeeded, self->lastKey, true));
+		}
+		return Void();
+	}
+
+	ACTOR static Future<bool> finishCurTenantBlockStartNewIfNeeded(EncryptedRangeFileWriter* self,
+	                                                               Key k,
+	                                                               Value v,
+	                                                               bool writeValue) {
+		// Don't want to start a new block if the current key or previous key is empty
+		if (self->lastKey.size() == 0 || k.size() == 0) {
+			return false;
+		}
+		state std::pair<int64_t, TenantName> curKeyTenantInfo = wait(getTenantIdName(k, self->tenantCache));
+		state std::pair<int64_t, TenantName> prevKeyTenantInfo =
+		    wait(getTenantIdName(self->lastKey, self->tenantCache));
+		// crossing tenant boundaries so finish the current block using only the tenant prefix of the new key
+		if (curKeyTenantInfo.first != prevKeyTenantInfo.first) {
+			CODE_PROBE(true, "crossed tenant boundaries");
+			state KeyRef endKey = k;
+			// If we are crossing a boundary with a key that has a tenant prefix then truncate it
+			if (curKeyTenantInfo.first != SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID &&
+			    curKeyTenantInfo.first != FDB_DEFAULT_ENCRYPT_DOMAIN_ID) {
+				endKey = StringRef(k.begin(), 8);
+			}
+			state ValueRef newValue = StringRef();
+			self->lastKey = k;
+			self->lastValue = v;
+			appendStringRefWithLenToBuffer(self, &endKey);
+			appendStringRefWithLenToBuffer(self, &newValue);
+			wait(newBlock(self, 0, endKey, writeValue));
+			wait(updateEncryptionKeys(self, self->lastKey, self->tenantCache));
+			return true;
+		}
+		return false;
+	}
+
+	// Start a new block if needed, then write the key and value
+	ACTOR static Future<Void> writeKV_impl(EncryptedRangeFileWriter* self, Key k, Value v) {
+		if (!self->cipherKeys.headerCipherKey.isValid() || !self->cipherKeys.textCipherKey.isValid()) {
+			wait(updateEncryptionKeys(self, k, self->tenantCache));
+		}
+		state int toWrite = sizeof(int32_t) + k.size() + sizeof(int32_t) + v.size();
+		wait(newBlockIfNeeded(self, toWrite));
+		bool createdNewBlock = wait(finishCurTenantBlockStartNewIfNeeded(self, k, v, true));
+		if (createdNewBlock) {
+			return Void();
+		}
+		appendStringRefWithLenToBuffer(self, &k);
+		appendStringRefWithLenToBuffer(self, &v);
+		self->lastKey = k;
+		self->lastValue = v;
+		return Void();
+	}
+
+	Future<Void> writeKV(Key k, Value v) { return writeKV_impl(this, k, v); }
+
+	// Write begin key or end key.
+	ACTOR static Future<Void> writeKey_impl(EncryptedRangeFileWriter* self, Key k) {
+		if (k.size() > 0 &&
+		    (!self->cipherKeys.headerCipherKey.isValid() || !self->cipherKeys.textCipherKey.isValid())) {
```

Comment:
ditto?

Reply:
This gets called when we write the first and last key of a file so its possible we haven't fetched the encryption keys at this point

Reply:
Interesting. What is the significance of `first` and/or `last` key to be empty? Also, `first` key to be empty seems like a bug, why would we write `empty key` and am assuming value also is `empty`; hence, technically a NOP? 

Last key I thought would be at the least: `tenantPrefix` if tenant is available, else, CommonPrefix between last-key of the block and first key of the next-block?

Reply:
Also, it is better to document such nuances, as it might be fresh in our minds at the moment, but, then sometime in future hard to reason about :)

Reply:
Yes I have seen the first or last key is sometimes empty. Why it happens I am not sure, will require more digging. I can document it though

Reply:
The bigger question I have is: empty first/last key is acceptable OR it may be due to an existing bug?

Reply:
Yeah i think this would be a better question for @sfc-gh-satherton, as I've seen this been done in tests. I think he's OOO this week so maybe i'll add as a TODO to investigate this